File: electron/src/session-host.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// THE SESSION HOST — the utilityProcess's whole lifecycle, and nothing else.
//
// One process per SESSION, not per app. The dedicated server runs several,
// and one process per session is what makes that free; it is also what keeps
// the engine's 36 process-global mutable bindings — the `BALANCE` tuning
// object, the six flags in `engine/game/flags.ts`, every `activeXDefs` catalog
// `registerDefs` swaps when a mod loads — from being shared between two runs
// that disagree about them.
//
// `utilityProcess` rather than the main process, and rather than a
// `BrowserWindow`: it is a real Node child with an IPC channel and, crucially,
// the ability to be handed a `MessagePort` that reaches the RENDERER directly.
// That port is the whole reason snapshots do not cost the main process
// anything — see `net.ts` for the split between control and game traffic.
//
// **A crashed session must look like a crashed session.** The exit handler
// fires for a clean stop and for a segfault alike, so the reason is recorded
// BEFORE the kill and read back in the handler; without that, a server that
// died mid-run and one the player asked to stop are indistinguishable, and the
// HOST screen would say "stopped" over a crash.

import {
  utilityProcess,
  type MessagePortMain,
  type UtilityProcess,
} from "electron";

import { output } from "./output";
import { serverEntryPath } from "./resources";

/** Where a socket ACTUALLY ended up. Never the port that was requested — see
 * `server/net/udp.ts` for why that distinction is a rule rather than a detail. */
export type ServerBound = { address: string; port: number };

/** What the ROUTER row on the HOST screen reads. Mirrors `MappingState` in
 * `server/net/upnp.ts`. */
export type ServerMapping =
  | { status: "idle" }
  | { status: "mapping" }
  | {
      status: "mapped";
      method: "nat-pmp" | "upnp";
      externalAddress: string | null;
      externalPort: number;
    }
  | { status: "failed"; detail: string };

/** One seat, as everybody else may see it. Mirrors `RosterEntry` in
 * `server/wire/protocol.ts`. */
export type ServerRosterEntry = {
  slot: number;
  name: string;
  playing: boolean;
  ping: number;
  rate: number;
};

/** A message down the control channel to the server. Mirrors `ControlMessage`
 * in `server/main.ts` — keep the two in step. */
export type ServerControl =
  | {
      kind: "start";
      params: unknown;
      mods?: string[];
      /** The catalog overrides those mods registered, opaque like
       * `adopt` — the session registers them before it builds. */
      modDefs?: unknown;
      password?: string;
      maxClients?: number;
      /** Seats to fill with the session's own AUTOPILOT heroes. A session fact
       * like `maxClients` beside it — never on the params, which describe the
       * run itself. */
      bots?: number;
      /** A run the session should ADOPT rather than build. Opaque here — this
       * process never learns what a run is, which is the point of the two
       * channels (see net.ts). */
      adopt?: unknown;
      /** Whether the hub may admit peers over a transport other than the
       * Steam relay. The shell decides it from what the build permits. */
      allowDirect?: boolean;
    }
  | { kind: "stop" }
  | { kind: "status" }
  | {
      kind: "listen";
      port?: number;
      udp?: boolean;
      steam?: boolean;
      /** Whether the session may ask the router to forward the bound port. */
      map?: boolean;
    }
  /** The other role: JOIN somebody else's session instead of hosting one. The
   * same fork, the same port to the renderer — see `server/main.ts`. */
  | {
      kind: "connect";
      address?: string;
      peer?: string;
      name: string;
      password?: string;
      mods?: string[];
      hardcore?: boolean;
      /** The hero this player brings, opaque here like `adopt`. */
      loadout?: unknown;
    }
  | { kind: "peer"; from: string; data: number[] }
  | { kind: "peer-lost"; from: string; reason: string };

/** A message back up it. Mirrors `ControlReply` in `server/main.ts`. */
export type ServerReply =
  | { kind: "ready"; protocol: number }
  | { kind: "started"; levelId: string }
  | {
      kind: "status";
      tick: number;
      phase: string;
      enemies: number;
      clients: number;
      bound: ServerBound | null;
      mapping: ServerMapping;
      roster: ServerRosterEntry[];
    }
  | {
      kind: "listening";
      bound: ServerBound | null;
      steam: boolean;
      protocol: number;
      build: string;
      detail?: string;
    }
  | { kind: "connected"; ok: boolean; reason?: string; detail?: string }
  | {
      kind: "peer-send";
      to: string;
      data: number[];
      mode: "reliable" | "unreliable";
    }
  | { kind: "invite" }
  | { kind: "log"; line: string }
  | { kind: "stopped"; reason: string }
  | { kind: "error"; detail: string };

/**
 * The replies nothing is ever WAITING on.
 *
 * The bridge above matches replies to requests by ORDER — the server answers in
 * order, so a queue is enough and no correlation id has to cross. That only
 * holds if unsolicited messages are excluded, and getting this list wrong is
 * silent: a log line would settle whatever request happened to be in flight,
 * and the HOST screen would report a refusal it was never sent.
 */
export const UNSOLICITED: ReadonlySet<ServerReply["kind"]> = new Set([
  "ready",
  "peer-send",
  "invite",
  "log",
]);

export type SessionHostOptions = {
  /** Every reply the server sends, in order. */
  onReply: (reply: ServerReply) => void;
  /** The process ended. `expected` is false for a crash. */
  onExit: (code: number, expected: boolean) => void;
  /** Fork a child. Injected so the lifecycle is testable without an Electron
   * runtime — the desktop check job deliberately installs no 100 MB binary
   * (see electron/tests/), and a supervisor that can only be tested by
   * launching the app is a supervisor that is never tested. */
  fork?: ForkFn;
};

/** The one thing this module needs from Electron, narrowed to what it uses. */
export type ForkFn = (entry: string) => UtilityProcess;

export type SessionHost = {
  /** True between `start()` and the process exiting. */
  readonly running: boolean;
  start(): void;
  /** Send one control message. A no-op when nothing is running. */
  send(message: ServerControl): void;
  /** Hand the server the renderer's end of the snapshot channel. */
  givePort(port: MessagePortMain, message: ServerControl): void;
  /** Ask for an orderly shutdown, then kill if it does not come. */
  stop(): void;
};

/** How long a `stop` may take to be honoured before the process is killed.
 * Short: the server's own stop is synchronous, so anything past this is a
 * process that is no longer answering, and a host that will not quit is worse
 * than one that is killed. */
const SHUTDOWN_GRACE_MS = 2_000;

export function createSessionHost(options: SessionHostOptions): SessionHost {
  const fork: ForkFn = options.fork ?? ((entry) => utilityProcess.fork(entry));
  let child: UtilityProcess | null = null;
  let expectedExit = false;
  let killTimer: ReturnType<typeof setTimeout> | null = null;

  function clearKillTimer(): void {
    if (killTimer) clearTimeout(killTimer);
    killTimer = null;
  }

  return {
    get running() {
      return child !== null;
    },

    start() {
      if (child) return;
      expectedExit = false;
      const entry = serverEntryPath();
      output.info(`session server: forking ${entry}`);
      child = fork(entry);
      child.on("message", (message: unknown) => {
        options.onReply(message as ServerReply);
      });
      child.on("exit", (code: number) => {
        clearKillTimer();
        const wasExpected = expectedExit;
        child = null;
        if (!wasExpected) {
          output.warn(`session server exited unexpectedly (code ${code})`);
        }
        options.onExit(code, wasExpected);
      });
    },

    send(message) {
      child?.postMessage(message);
    },

    givePort(port, message) {
      // The port travels WITH the message that starts the session, so the
      // server has it before there is anything to send down it. The server
      // tolerates either order (it holds an early port), but arriving together
      // is what makes the first snapshot the first thing the client sees.
      child?.postMessage(message, [port]);
    },

    stop() {
      if (!child) return;
      expectedExit = true;
      child.postMessage({ kind: "stop" } satisfies ServerControl);
      clearKillTimer();
      killTimer = setTimeout(() => {
        killTimer = null;
        if (!child) return;
        output.warn("session server did not stop; killing it");
        child.kill();
      }, SHUTDOWN_GRACE_MS);
    },
  };
}

File: tauri/shell/src/session_host.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! THE SESSION SIDECAR'S PROTOCOL AND LIFECYCLE RULES — the peer of
//! `electron/src/session-host.ts`, minus the process itself.
//!
//! One process per SESSION, not per app, for the same three reasons that file
//! gives: a 60 Hz simulation must not compete with the shell, the engine holds
//! process-global mutable bindings that two runs would fight over, and it makes
//! the host's own renderer just another client.
//!
//! **WHERE THE TWO SHELLS DIVERGE IS THE PIPE, AND ONLY THE PIPE.** Electron
//! forks a `utilityProcess`, which is a real Node child with an IPC channel and
//! the ability to be handed a `MessagePort` that reaches the renderer. Tauri can
//! spawn a child and nothing more, so:
//!
//! | | Electron | Here |
//! | --- | --- | --- |
//! | the child | `utilityProcess.fork` | `std::process::Command` on a Node runtime |
//! | control | the Node IPC channel | the child's STDIO, newline-delimited JSON |
//! | snapshots | a transferred `MessagePort` | a loopback socket the PAGE opens |
//! | reaping | `before-quit` kills it | stdin's EOF, which the child watches |
//!
//! `server/shell-host.ts` is the far end of all four rows and carries the
//! argument for each. What is HERE is everything that can be decided without a
//! process: what a reply is, which replies nobody is waiting on, how long a
//! stop may take, and what the child's first line means.
//!
//! **A CRASHED SESSION MUST LOOK LIKE A CRASHED SESSION**, which is the one
//! rule that survived the move unchanged: the reason is recorded BEFORE the kill
//! and read back in the exit handler, because otherwise a server that died
//! mid-run and one the player asked to stop are indistinguishable and the HOST
//! screen says "stopped" over a crash.

use serde_json::Value;

/// How long a `stop` may take to be honoured before the process is killed.
///
/// Short, exactly as on the Electron side: the server's own stop is
/// synchronous, so anything past this is a process that is no longer answering,
/// and a host that will not quit is worse than one that is killed.
pub const SHUTDOWN_GRACE_MS: u64 = 2_000;

/// How long the child may take to report its snapshot endpoint before the
/// launch is called a failure.
///
/// It is a Node process starting and binding a loopback socket — milliseconds
/// on a warm disk, seconds on a cold one, and never the tens of seconds a
/// LEVEL takes to build. The HOST screen's own timeout is longer; this one
/// exists so a runtime that is missing entirely produces a refusal here rather
/// than a fifteen-second silence.
pub const READY_TIMEOUT_MS: u64 = 20_000;

/// Where the page must open the snapshot channel.
///
/// Minted by the CHILD rather than by the shell, and reported on its first
/// line: the port is ephemeral, so only the process that bound it knows it, and
/// a shell that picked one first would be racing whatever else on the machine
/// wanted it. The token comes the same way for the same reason — one mint, one
/// owner.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SnapshotEndpoint {
    /// The loopback port the child bound.
    pub port: u16,
    /// The secret the upgrade has to carry.
    pub token: String,
    /// The one path an upgrade may name.
    pub path: String,
}

impl SnapshotEndpoint {
    /// The URL the page connects to.
    ///
    /// `127.0.0.1` rather than `localhost`, deliberately: the name resolves to
    /// two addresses on a dual-stack machine and the child bound one of them,
    /// so a page that got the other would report a connection refused that
    /// nothing in either process is wrong about.
    pub fn url(&self) -> String {
        format!(
            "ws://127.0.0.1:{}{}?token={}",
            self.port, self.path, self.token
        )
    }
}

/// A reply from the session process, as much of it as the shell reads.
///
/// Mirrors `ControlReply` in `server/main.ts` — keep the two in step. Anything
/// this shell does not know is [`ServerReply::Unknown`] rather than a parse
/// failure, because a newer server talking to an older shell is a situation
/// that should degrade rather than crash.
#[derive(Debug, Clone, PartialEq)]
pub enum ServerReply {
    /// The process is up. Carries the snapshot endpoint in the sidecar entry.
    Ready {
        /// The wire protocol the session speaks.
        protocol: u32,
        /// Where the page opens its channel, absent only if the child was
        /// somehow started without the sidecar flag.
        snapshot: Option<SnapshotEndpoint>,
    },
    /// A session was built.
    Started {
        /// The level it built.
        level_id: String,
    },
    /// One poll of the HOST screen.
    Status(Value),
    /// The doors, as they actually opened.
    Listening {
        /// The whole reply, for the page — the shell reads three fields off it
        /// and forwards the rest untouched.
        detail: Value,
    },
    /// A join settled.
    Connected {
        /// Whether the door opened.
        ok: bool,
        /// The host's own refusal, when it did not.
        reason: Option<String>,
        /// More of it, when there is more.
        detail: Option<String>,
    },
    /// One packet for the shell to put on the Steam P2P queue.
    PeerSend {
        /// The peer's Steam id, as a decimal string.
        to: String,
        /// The bytes.
        data: Vec<u8>,
        /// Whether it must arrive.
        reliable: bool,
    },
    /// The session asked for the platform's invite panel.
    Invite,
    /// One line for the shell's own log.
    Log {
        /// What to write down.
        line: String,
    },
    /// The session ended.
    Stopped,
    /// Something went wrong inside the session.
    Error {
        /// What it said.
        detail: String,
    },
    /// A reply this shell has no case for.
    Unknown {
        /// Its `kind`, for the log.
        kind: String,
    },
}

impl ServerReply {
    /// The `kind` string this reply arrived as — what a waiter is matched on.
    pub fn kind(&self) -> &str {
        match self {
            ServerReply::Ready { .. } => "ready",
            ServerReply::Started { .. } => "started",
            ServerReply::Status(_) => "status",
            ServerReply::Listening { .. } => "listening",
            ServerReply::Connected { .. } => "connected",
            ServerReply::PeerSend { .. } => "peer-send",
            ServerReply::Invite => "invite",
            ServerReply::Log { .. } => "log",
            ServerReply::Stopped => "stopped",
            ServerReply::Error { .. } => "error",
            ServerReply::Unknown { kind } => kind,
        }
    }

    /// Is this one nothing is ever WAITING on?
    ///
    /// The bridge matches replies to requests by ORDER — the server answers in
    /// order, so a queue is enough and no correlation id has to cross, which
    /// keeps `server/main.ts` free of request bookkeeping the dedicated server
    /// would also have to carry. That only holds if unsolicited messages are
    /// excluded, and getting this list wrong is SILENT: a log line would settle
    /// whatever request happened to be in flight, and the HOST screen would
    /// report a refusal it was never sent.
    pub fn unsolicited(&self) -> bool {
        matches!(
            self,
            ServerReply::Ready { .. }
                | ServerReply::PeerSend { .. }
                | ServerReply::Invite
                | ServerReply::Log { .. }
        )
    }
}

/// Read one line of the child's stdout.
///
/// `None` for anything that is not a JSON object with a `kind` — which includes
/// a blank line, and includes whatever a dependency decided to print. The
/// child's stdout is a protocol channel and this shell treats stray output as
/// noise rather than as a reason to stop reading, because a session killed by
/// somebody else's `console.log` would be a spectacular way to lose a run.
pub fn parse_reply(line: &str) -> Option<ServerReply> {
    let value: Value = serde_json::from_str(line.trim()).ok()?;
    let kind = value.get("kind")?.as_str()?;
    let text = |field: &str| value.get(field).and_then(Value::as_str).map(str::to_string);
    Some(match kind {
        "ready" => ServerReply::Ready {
            protocol: value.get("protocol").and_then(Value::as_u64).unwrap_or(0) as u32,
            snapshot: parse_snapshot(value.get("snapshot")),
        },
        "started" => ServerReply::Started {
            level_id: text("levelId").unwrap_or_default(),
        },
        "status" => ServerReply::Status(value),
        "listening" => ServerReply::Listening { detail: value },
        "connected" => ServerReply::Connected {
            ok: value.get("ok").and_then(Value::as_bool).unwrap_or(false),
            reason: text("reason"),
            detail: text("detail"),
        },
        "peer-send" => ServerReply::PeerSend {
            to: text("to").unwrap_or_default(),
            data: value
                .get("data")
                .and_then(Value::as_array)
                .map(|bytes| {
                    bytes
                        .iter()
                        .filter_map(Value::as_u64)
                        .map(|byte| byte as u8)
                        .collect()
                })
                .unwrap_or_default(),
            // Anything that is not explicitly reliable is not: the relay's
            // unreliable path is the one that carries snapshots, and a
            // mis-defaulted mode would put every frame through Valve's
            // retransmit queue.
            reliable: text("mode").as_deref() == Some("reliable"),
        },
        "invite" => ServerReply::Invite,
        "log" => ServerReply::Log {
            line: text("line").unwrap_or_default(),
        },
        "stopped" => ServerReply::Stopped,
        "error" => ServerReply::Error {
            detail: text("detail").unwrap_or_default(),
        },
        other => ServerReply::Unknown {
            kind: other.to_string(),
        },
    })
}

fn parse_snapshot(value: Option<&Value>) -> Option<SnapshotEndpoint> {
    let value = value?;
    let port = value.get("port").and_then(Value::as_u64)?;
    // Port 0 is what an unbound listener reports, and a page told to connect to
    // it would get a connection refused with nothing to say about why.
    if port == 0 || port > u64::from(u16::MAX) {
        return None;
    }
    let token = value.get("token")?.as_str()?.to_string();
    if token.is_empty() {
        return None;
    }
    Some(SnapshotEndpoint {
        port: port as u16,
        token,
        path: value
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or("/snapshot")
            .to_string(),
    })
}

/// The line the shell writes for a process that ended.
///
/// A sentence rather than a code: the reader is somebody holding a launch log
/// and wondering why the HOST screen went blank, and the difference between the
/// two cases is the whole answer.
pub fn describe_exit(code: Option<i32>, expected: bool) -> Option<String> {
    if expected {
        return None;
    }
    Some(match code {
        Some(code) => format!("session server exited unexpectedly (code {code})"),
        None => "session server was killed".to_string(),
    })
}
